### PR TITLE
ssl: Allow deletion of contexts at runtime

### PR DESCRIPTION
**Pull request by mattklein123**

### CODE REVIEW

**Comment by mattklein123** [comment]
@lyft/network-team

File: source/common/ssl/context_manager_impl.cc [hunk 1]
```diff
@@ -1,37 +1,49 @@
...
-                                                               ContextConfig& config) {
-  Ssl::ServerContext* context = new ServerContextImpl(scope, config, runtime_);
-  contexts_.emplace_back(context);
-  return *context;
+ClientContextPtr ContextManagerImpl::createSslClientContext(Stats::Scope& scope,
+                                                            ContextConfig& config) {
+
+  ClientContextPtr context(new ClientContextImpl(*this, scope, config));
+  std::unique_lock<std::mutex> lock(contexts_lock_);
+  contexts_.emplace_back(context.get());
```
Comment (wgallagher) [inline source/common/ssl/context_manager_impl.cc]: is this correct?  Won't the ClientContextPtr going out of scope delete the new ClientContextImpl?

File: source/common/upstream/upstream_impl.cc [hunk 2]
```diff
@@ -67,7 +67,7 @@ ClusterInfoImpl::ClusterInfoImpl(const Json::Object& config, Runtime::Loader& ru
   ssl_ctx_ = nullptr;
   if (config.hasObject("ssl_context")) {
     Ssl::ContextConfigImpl context_config(*config.getObject("ssl_context"));
-    ssl_ctx_ = &ssl_context_manager.createSslClientContext(stats_scope_, context_config);
+    ssl_ctx_ = ssl_context_manager.createSslClientContext(stats_scope_, context_config);
```
Comment (wgallagher) [inline source/common/upstream/upstream_impl.cc]: where is the call to ssl_context_manager.release() for this?

Comment (mattklein123) [inline source/common/upstream/upstream_impl.cc]: Inside the destructor of ContextImpl. 

FWIW, I'm not crazy about this solution, but I can't think of a better way. The caller is the only owner of the context, and the manager has a weak reference. It's cleaned up in destructor.

**Pull request merged**